// @veramo/key-manager/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryPrivateKeyStore = exports.MemoryKeyStore = exports.AbstractSecretBox = exports.AbstractPrivateKeyStore = exports.AbstractKeyStore = exports.AbstractKeyManagementSystem = exports.KeyManager = void 0;
/**
 * Provides a {@link @veramo/key-manager#KeyManager | plugin} for the {@link @veramo/core#Agent}
 * that implements {@link @veramo/core#IKeyManager} interface
 *
 * @packageDocumentation
 */
var key_manager_1 = require("./key-manager");
Object.defineProperty(exports, "KeyManager", { enumerable: true, get: function () { return key_manager_1.KeyManager; } });
var abstract_key_management_system_1 = require("./abstract-key-management-system");
Object.defineProperty(exports, "AbstractKeyManagementSystem", { enumerable: true, get: function () { return abstract_key_management_system_1.AbstractKeyManagementSystem; } });
var abstract_key_store_1 = require("./abstract-key-store");
Object.defineProperty(exports, "AbstractKeyStore", { enumerable: true, get: function () { return abstract_key_store_1.AbstractKeyStore; } });
var abstract_private_key_store_1 = require("./abstract-private-key-store");
Object.defineProperty(exports, "AbstractPrivateKeyStore", { enumerable: true, get: function () { return abstract_private_key_store_1.AbstractPrivateKeyStore; } });
var abstract_secret_box_1 = require("./abstract-secret-box");
Object.defineProperty(exports, "AbstractSecretBox", { enumerable: true, get: function () { return abstract_secret_box_1.AbstractSecretBox; } });
var memory_key_store_1 = require("./memory-key-store");
Object.defineProperty(exports, "MemoryKeyStore", { enumerable: true, get: function () { return memory_key_store_1.MemoryKeyStore; } });
Object.defineProperty(exports, "MemoryPrivateKeyStore", { enumerable: true, get: function () { return memory_key_store_1.MemoryPrivateKeyStore; } });
__exportStar(require("./types"), exports);
