// @veramo/data-store/build/index.js
"use strict";
/**
 * {@link https://typeorm.io/ | TypeORM } backed plugins. {@link @veramo/core#Agent}
 * {@link @veramo/data-store#DataStore | plugin} that implements {@link @veramo/core#IDataStore } interface.
 * {@link @veramo/core#Agent} {@link @veramo/data-store#DataStoreORM | plugin} that implements
 * {@link @veramo/core#IDataStoreORM} interface. Provides {@link @veramo/data-store#KeyStore} for
 * {@link @veramo/key-manager#KeyManager} and {@link @veramo/data-store#DIDStore} for
 * {@link @veramo/did-manager#DIDManager}
 *
 * @packageDocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrations = exports.PreMigrationKey = exports.PrivateKey = exports.Service = exports.Presentation = exports.Credential = exports.Claim = exports.Message = exports.Identifier = exports.Key = exports.Entities = exports.DataStoreDiscoveryProvider = exports.DataStoreORM = exports.DataStore = exports.PrivateKeyStore = exports.KeyStore = exports.DIDStore = void 0;
var did_store_1 = require("./identifier/did-store");
Object.defineProperty(exports, "DIDStore", { enumerable: true, get: function () { return did_store_1.DIDStore; } });
var key_store_1 = require("./identifier/key-store");
Object.defineProperty(exports, "KeyStore", { enumerable: true, get: function () { return key_store_1.KeyStore; } });
var private_key_store_1 = require("./identifier/private-key-store");
Object.defineProperty(exports, "PrivateKeyStore", { enumerable: true, get: function () { return private_key_store_1.PrivateKeyStore; } });
var data_store_1 = require("./data-store");
Object.defineProperty(exports, "DataStore", { enumerable: true, get: function () { return data_store_1.DataStore; } });
var data_store_orm_1 = require("./data-store-orm");
Object.defineProperty(exports, "DataStoreORM", { enumerable: true, get: function () { return data_store_orm_1.DataStoreORM; } });
var did_discovery_provider_1 = require("./did-discovery-provider");
Object.defineProperty(exports, "DataStoreDiscoveryProvider", { enumerable: true, get: function () { return did_discovery_provider_1.DataStoreDiscoveryProvider; } });
const key_1 = require("./entities/key");
Object.defineProperty(exports, "Key", { enumerable: true, get: function () { return key_1.Key; } });
const identifier_1 = require("./entities/identifier");
Object.defineProperty(exports, "Identifier", { enumerable: true, get: function () { return identifier_1.Identifier; } });
const claim_1 = require("./entities/claim");
Object.defineProperty(exports, "Claim", { enumerable: true, get: function () { return claim_1.Claim; } });
const credential_1 = require("./entities/credential");
Object.defineProperty(exports, "Credential", { enumerable: true, get: function () { return credential_1.Credential; } });
const presentation_1 = require("./entities/presentation");
Object.defineProperty(exports, "Presentation", { enumerable: true, get: function () { return presentation_1.Presentation; } });
const service_1 = require("./entities/service");
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return service_1.Service; } });
const message_1 = require("./entities/message");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return message_1.Message; } });
const private_key_1 = require("./entities/private-key");
Object.defineProperty(exports, "PrivateKey", { enumerable: true, get: function () { return private_key_1.PrivateKey; } });
const PreMigrationEntities_1 = require("./entities/PreMigrationEntities");
Object.defineProperty(exports, "PreMigrationKey", { enumerable: true, get: function () { return PreMigrationEntities_1.PreMigrationKey; } });
/**
 * The TypeORM entities used by this package.
 *
 * This array SHOULD be used when creating a TypeORM connection.
 *
 * @public
 */
exports.Entities = [
    key_1.Key,
    identifier_1.Identifier,
    message_1.Message,
    claim_1.Claim,
    credential_1.Credential,
    presentation_1.Presentation,
    service_1.Service,
    private_key_1.PrivateKey,
    PreMigrationEntities_1.PreMigrationKey,
];
var migrations_1 = require("./migrations");
Object.defineProperty(exports, "migrations", { enumerable: true, get: function () { return migrations_1.migrations; } });
