// @veramo/did-resolver/build/resolver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DIDResolverPlugin = void 0;
const core_1 = require("@veramo/core");
const utils_1 = require("@veramo/utils");
const did_resolver_1 = require("did-resolver");
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('veramo:resolver');
/**
 * A Veramo Plugin that enables users to resolve DID documents.
 *
 * This plugin is used automatically by plugins that create or verify Verifiable Credentials or Presentations or when
 * working with DIDComm
 *
 * @public
 */
class DIDResolverPlugin {
    constructor(options) {
        this.schema = core_1.schema.IResolver;
        const { resolver } = options, resolverMap = __rest(options, ["resolver"]);
        if ((0, utils_1.isDefined)(resolver)) {
            this.didResolver = resolver;
        }
        else if (Object.keys(resolverMap).length > 0) {
            this.didResolver = new did_resolver_1.Resolver(resolverMap);
        }
        else {
            throw Error('invalid_setup: The DIDResolverPlugin must be initialized with a Resolvable or a map of methods to DIDResolver implementations');
        }
        this.methods = {
            resolveDid: this.resolveDid.bind(this),
            getDIDComponentById: this.getDIDComponentById.bind(this),
        };
    }
    /** {@inheritDoc @veramo/core#IResolver.resolveDid} */
    resolveDid({ didUrl, options, }) {
        return __awaiter(this, void 0, void 0, function* () {
            debug('Resolving %s', didUrl);
            const resolverOptions = Object.assign({ accept: 'application/did+ld+json' }, options);
            // ensure the required fields are present, even if the resolver is not compliant
            const cannedResponse = {
                didDocumentMetadata: {},
                didResolutionMetadata: {},
                didDocument: null,
            };
            const resolution = yield this.didResolver.resolve(didUrl, resolverOptions);
            return Object.assign(Object.assign({}, cannedResponse), resolution);
        });
    }
    /** {@inheritDoc @veramo/core#IResolver.getDIDComponentById} */
    getDIDComponentById({ didDocument, didUrl, section, }) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            debug('Resolving %s', didUrl);
            const did = ((_a = (0, did_resolver_1.parse)(didUrl)) === null || _a === void 0 ? void 0 : _a.did) || didDocument.id;
            const doc = didDocument;
            const mainSections = [...(doc.verificationMethod || []), ...(doc.publicKey || []), ...(doc.service || [])];
            const subsection = section ? [...(doc[section] || [])] : mainSections;
            let result = subsection.find((item) => {
                if (typeof item === 'string') {
                    return item === didUrl || `${did}${item}` === didUrl;
                }
                else {
                    return item.id === didUrl || `${did}${item.id}` === didUrl;
                }
            });
            if (typeof result === 'string') {
                result = mainSections.find((item) => item.id === didUrl || `${did}${item.id}` === didUrl);
            }
            if (!result) {
                const err = `not_found: DID document fragment (${didUrl}) could not be located.`;
                debug(err);
                throw new Error(err);
            }
            else if (result.id.startsWith('#')) {
                // fix did documents that use only the fragment part as key ID
                result.id = `${did}${result.id}`;
            }
            return result;
        });
    }
}
exports.DIDResolverPlugin = DIDResolverPlugin;
