// @tuum-tech/did-provider-pkh/build/hedera/hedera-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleHederaClientImpl = void 0;
const sdk_1 = require("@hashgraph/sdk");
const create_account_1 = require("./create-account");
/* eslint-disable */
class SimpleHederaClientImpl {
    constructor(client, privateKey) {
        this._client = client;
        this._privateKey = privateKey;
    }
    getPrivateKey() {
        return this._privateKey;
    }
    getPublicKey() {
        /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
        return this._client.operatorPublicKey;
    }
    getAccountInfo(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            // Create the account info query
            const query = new sdk_1.AccountInfoQuery().setAccountId(accountId);
            // Sign with client operator private key and submit the query to a Hedera network
            const accountInfo = yield query.execute(this._client);
            return accountInfo;
        });
    }
    getAccountId() {
        /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
        return this._client.operatorAccountId;
    }
    createAccount(options) {
        return (0, create_account_1.createAccount)(this._client, options);
    }
}
exports.SimpleHederaClientImpl = SimpleHederaClientImpl;
