// @veramo/credential-ld/build/suites/EcdsaSecp256k1RecoverySignature2020.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VeramoEcdsaSecp256k1RecoverySignature2020 = void 0;
const ld_suites_1 = require("../ld-suites");
const lds_ecdsa_secp256k1_recovery2020_1 = require("@veramo-community/lds-ecdsa-secp256k1-recovery2020");
const u8a = __importStar(require("uint8arrays"));
const utils_1 = require("@veramo/utils");
/**
 * Veramo wrapper for the EcdsaSecp256k1RecoverySignature2020 suite by Transmute Industries
 *
 * @alpha This API is experimental and is very likely to change or disappear in future releases without notice.
 */
class VeramoEcdsaSecp256k1RecoverySignature2020 extends ld_suites_1.VeramoLdSignature {
    getSupportedVerificationType() {
        return 'EcdsaSecp256k1RecoveryMethod2020';
    }
    getSupportedVeramoKeyType() {
        return 'Secp256k1';
    }
    getSuiteForSigning(key, did, verifiableMethodId, context) {
        const controller = did;
        const signer = {
            //returns a JWS detached
            sign: (args) => __awaiter(this, void 0, void 0, function* () {
                const header = {
                    alg: 'ES256K-R',
                    b64: false,
                    crit: ['b64'],
                };
                const headerString = (0, utils_1.encodeJoseBlob)(header);
                const messageBuffer = u8a.concat([u8a.fromString(`${headerString}.`, 'utf-8'), args.data]);
                const messageString = u8a.toString(messageBuffer, 'base64');
                const signature = yield context.agent.keyManagerSign({
                    keyRef: key.kid,
                    algorithm: 'ES256K-R',
                    data: messageString,
                    encoding: 'base64',
                });
                return `${headerString}..${signature}`;
            }),
        };
        const suite = new lds_ecdsa_secp256k1_recovery2020_1.EcdsaSecp256k1RecoverySignature2020({
            // signer,
            key: new lds_ecdsa_secp256k1_recovery2020_1.EcdsaSecp256k1RecoveryMethod2020({
                publicKeyHex: key.publicKeyHex,
                signer: () => signer,
                type: this.getSupportedVerificationType(),
                controller,
                id: verifiableMethodId,
            }),
        });
        suite.ensureSuiteContext = ({ document }) => {
            document['@context'] = [...(0, utils_1.asArray)(document['@context'] || []), this.getContext()];
        };
        return suite;
    }
    getSuiteForVerification() {
        return new lds_ecdsa_secp256k1_recovery2020_1.EcdsaSecp256k1RecoverySignature2020();
    }
    preSigningCredModification(credential) {
    }
    preDidResolutionModification(didUrl, didDoc) {
        var _a, _b;
        //    did:ethr
        const idx = ((_a = didDoc['@context']) === null || _a === void 0 ? void 0 : _a.indexOf('https://identity.foundation/EcdsaSecp256k1RecoverySignature2020/lds-ecdsa-secp256k1-recovery2020-0.0.jsonld')) || -1;
        if (Array.isArray(didDoc['@context']) && idx !== -1) {
            didDoc['@context'][idx] = this.getContext();
        }
        if (didUrl.toLowerCase().startsWith('did:ethr')) {
            //EcdsaSecp256k1RecoveryMethod2020 does not support older format blockchainAccountId
            (_b = didDoc.verificationMethod) === null || _b === void 0 ? void 0 : _b.forEach((x) => {
                if (x.blockchainAccountId) {
                    if (x.blockchainAccountId.lastIndexOf('@eip155:') !== -1) {
                        const [address, chain] = x.blockchainAccountId.split("@eip155:");
                        x.blockchainAccountId = `eip155:${chain}:${address}`;
                    }
                }
            });
        }
    }
    getContext() {
        return 'https://w3id.org/security/suites/secp256k1recovery-2020/v2';
    }
}
exports.VeramoEcdsaSecp256k1RecoverySignature2020 = VeramoEcdsaSecp256k1RecoverySignature2020;
