// @veramo/credential-ld/build/suites/Ed25519Signature2020.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VeramoEd25519Signature2020 = void 0;
const ld_suites_1 = require("../ld-suites");
const u8a = __importStar(require("uint8arrays"));
const ed25519_signature_2020_1 = require("@digitalcredentials/ed25519-signature-2020");
const ed25519_verification_key_2020_1 = require("@digitalcredentials/ed25519-verification-key-2020");
/**
 * Veramo wrapper for the Ed25519Signature2020 suite by digitalcredentials
 *
 * @alpha This API is experimental and is very likely to change or disappear in future releases without notice.
 */
class VeramoEd25519Signature2020 extends ld_suites_1.VeramoLdSignature {
    constructor() {
        super(...arguments);
        this.MULTIBASE_BASE58BTC_PREFIX = 'z';
        this.MULTICODEC_PREFIX = [0xed, 0x01];
    }
    getSupportedVerificationType() {
        return 'Ed25519VerificationKey2020';
    }
    getSupportedVeramoKeyType() {
        return 'Ed25519';
    }
    getSuiteForSigning(key, issuerDid, verificationMethodId, context) {
        const controller = issuerDid;
        // DID Key ID
        let id = verificationMethodId;
        const signer = {
            // returns signatureBytes
            sign: (args) => __awaiter(this, void 0, void 0, function* () {
                const messageString = u8a.toString(args.data, 'base64');
                const signature = yield context.agent.keyManagerSign({
                    keyRef: key.kid,
                    data: messageString,
                    encoding: 'base64',
                });
                return u8a.fromString(signature);
            }),
        };
        const verificationKey = new ed25519_verification_key_2020_1.Ed25519VerificationKey2020({
            id,
            controller,
            publicKeyMultibase: this.preSigningKeyModification(u8a.fromString(key.publicKeyHex, 'hex')),
            signer: () => signer,
            type: this.getSupportedVerificationType(),
        });
        // overwrite the signer since we're not passing the private key
        verificationKey.signer = () => signer;
        return new ed25519_signature_2020_1.Ed25519Signature2020({
            key: verificationKey,
            signer: signer,
        });
    }
    getSuiteForVerification() {
        return new ed25519_signature_2020_1.Ed25519Signature2020();
    }
    preSigningCredModification(credential) {
        // nothing to do here
    }
    preDidResolutionModification(didUrl, didDoc) {
        // nothing to do here
    }
    preSigningKeyModification(key) {
        const modifiedKey = u8a.concat([this.MULTICODEC_PREFIX, key]);
        return `${this.MULTIBASE_BASE58BTC_PREFIX}${u8a.toString(modifiedKey, 'base58btc')}`;
    }
}
exports.VeramoEd25519Signature2020 = VeramoEd25519Signature2020;
